### PR TITLE
fix: Use random state in sample_pdf, Resolves #3116 

### LINKED ISSUE #3116: CATS does not save the model
[003-Cats-Tutorial.ipynb.zip](https://github.com/VowpalWabbit/vowpal_wabbit/files/6748242/003-Cats-Tutorial.ipynb.zip)
I train a multi-armed bandit model for a continuous action space.
After training, I save the model, but after unloading the model, the progress is not saved or the model does not predict on old examples.

`
vw.predict(ex)
`
Result:
`
(13.885909080505371, 0.40625)
`

```
vw.save("mab.model")

model = pyvw.vw("mab.model")

model.predict(ex)
```
Result:
`
0.0
`

If I use loading with `--quiet -i`, then a bug comes when predicting

```
RuntimeError: Continuous actions labels require the first word to be ca

```

#### Environment
VW 8.10.2 on MAC OS with the Python bindings

### PR DESCRIPTION
**Pull request by bassmang**

### CODE REVIEW

File: test/unit_test/explore_test.cc [hunk 1]
```diff
@@ -97,13 +97,13 @@ BOOST_AUTO_TEST_CASE(sample_continuous_action_statistical)
   float pdf_value;
   constexpr float range_min = .0f;
   constexpr float range_max = 100.0f;
-  uint64_t random_seed = 7791;
+  std::shared_ptr<rand_state> random_seed = std::make_shared<rand_state>(7791);
```
Comment (jackgerrits) [inline test/unit_test/explore_test.cc:100]:
```suggestion
  auto random_seed = std::make_shared<rand_state>(7791);
```

File: explore/explore_internal.h [hunk 2]
```diff
@@ -340,8 +341,8 @@ namespace exploration
   // 2) Does not normalize the scores (unlike sample_after_normalization)
   // 3) Scores need not add up to one.
   template <typename It>
-  int sample_scores(
-      uint64_t* p_seed, It scores_first, It scores_last, uint32_t& chosen_index, std::random_access_iterator_tag)
+  int sample_scores(std::shared_ptr<rand_state> p_seed, It scores_first, It scores_last, uint32_t& chosen_index,
```
Comment (jackgerrits) [inline explore/explore_internal.h:344]: This function is part of the explore library which is independent, and therefore cannot depend on `rand_state`. The inner random state should be retrieved before this is called, then it should be updated with the updated state once `sample_scores` is complete.

File: vowpalwabbit/sample_pdf.cc [hunk 3]
```diff
@@ -34,10 +34,10 @@ struct sample_pdf
   int learn(example& ec, experimental::api_status* status);
   int predict(example& ec, experimental::api_status* status);
 
-  void init(single_learner* p_base, uint64_t* p_random_seed);
+  void init(single_learner* p_base, std::shared_ptr<rand_state>&& random_state);
```
Comment (jackgerrits) [inline vowpalwabbit/sample_pdf.cc:37]: Why make this an rvalue? This should be a value which is moved inside the call

File: vowpalwabbit/sample_pdf.cc [hunk 4]
```diff
@@ -71,10 +71,10 @@ int sample_pdf::predict(example& ec, experimental::api_status*)
   return VW::experimental::error_code::success;
 }
 
-void sample_pdf::init(single_learner* p_base, uint64_t* p_random_seed)
+void sample_pdf::init(single_learner* p_base, std::shared_ptr<rand_state>&& random_state)
```
Comment (jackgerrits) [inline vowpalwabbit/sample_pdf.cc:74]:
```suggestion
void sample_pdf::init(single_learner* p_base, std::shared_ptr<rand_state> random_state)
```

File: vowpalwabbit/sample_pdf.cc [hunk 5]
```diff
@@ -71,10 +71,10 @@ int sample_pdf::predict(example& ec, experimental::api_status*)
...
 {
   _base = p_base;
-  _p_random_state = p_random_seed;
+  _p_random_state = random_state;
```
Comment (jackgerrits) [inline vowpalwabbit/sample_pdf.cc:77]:
```suggestion
  _p_random_state = std::move(random_state);
```

**Review by jackgerrits — CHANGES_REQUESTED** [review]

File: vowpalwabbit/sample_pdf.cc (hunk 5 above)
Comment (olgavrou) [reply vowpalwabbit/sample_pdf.cc:77]: Do we want to move the shared pointer here?

Comment (jackgerrits) [reply vowpalwabbit/sample_pdf.cc:77]: If the argument is by value then moving it here is what we want. If the arg is by ref then letting it copy here is good.

Comment (olgavrou) [reply vowpalwabbit/sample_pdf.cc:77]: yeah was wondering if it's safe but looks like all.`get_random_state()` returns a copy of the shared pointer, so passing it as an rvalue and moving it seems fine (and probably a tad faster)

Comment (jackgerrits) [reply vowpalwabbit/sample_pdf.cc:77]: Either way works

File: test/unit_test/explore_test.cc (hunk 1 above)
Comment (bassmang) [reply test/unit_test/explore_test.cc]: I reverted explore_internal.h so seed can be a uint64_t here again

File: explore/explore_internal.h (hunk 2 above)
Comment (bassmang) [reply explore/explore_internal.h]: Changed to get seed in sample_pdf before calling this

**Review by jackgerrits — APPROVED** [review]

**Review by olgavrou — APPROVED** [review]

**Pull request merged**